Added junit format

ekho opened the pull request:
This pull request allowes to use php-cs-fixer in CI

Depends on #1824

keradus commented:
> This pull request allowes to use php-cs-fixer in CI

?
It could be used without it as well

keradus commented:
(Note: I'm not against adding new format)

ekho commented:
I mean it allows to view results in CI UI without greping logs

keradus commented:
I guess you refer to some particular CI service.
No CI as Continuous Integration service in general.

ekho commented:
ex Atlassian Bamboo https://confluence.atlassian.com/bamboo/junit-parser-289277056.html

stof commented on src/Report/JunitReport.php:
```diff
@@ -0,0 +1,113 @@
+        $file,
+        $fixResult,
+        $dom,
+        isset($fixFiles[$file]) ? $fixFiles[$file] : null
```
the indentation is wrong here

ekho replied:
I'll fix it, thanks

ekho commented:
For Jenkins too 
https://www.safaribooksonline.com/library/view/jenkins-the-definitive/9781449311155/ch06s03.html

keradus commented on src/Report/JunitReport.php:
```diff
@@ -0,0 +1,113 @@
+
+        /** @var \DomElement $testsuite */
+        $testsuite = $testsuites->appendChild($dom->createElement('testsuite'));
+        $testsuite->setAttribute('name', 'PhpCsFixer'.($this->getApplication() ? '-'.$this->getApplication()->getVersion() : ''));
```
:-1: for appending version to report.
especially if only that one will have it.

ekho replied:
I got this idea from php code sniffer source :)

Maybe it's unnecessary

GrahamCampbell commented:
Needs rebasing. ;)

ekho commented:
@GrahamCampbell Of course. I will rewrite it after #1824 will be approved

keradus commented:
#1824 is merged, please rebase

keradus commented on src/Console/ConfigurationResolver.php:
```diff
@@ -482,7 +482,7 @@ function (FixerInterface $fixer) {
     protected function resolveFormat()
     {
-        static $formats = array('txt', 'xml', 'json');
+	static $formats = array('txt', 'xml', 'json', 'junit');
```
?

ekho commented:
@keradus rebased

keradus commented on src/Report/JunitReporter.php:
```diff
@@ -0,0 +1,145 @@
+        $testsuites = $dom->appendChild($dom->createElement('testsuites'));
+        /** @var \DomElement $testsuite */
+        $testsuite = $testsuites->appendChild($dom->createElement('testsuite'));
+        $testsuite->setAttribute('name', 'PhpCsFixer report');
```
`PHP CS Fixer report`

Anahkiasen commented:
Would love this, CircleCI also uses JUnit as format, currently doing `php-cs-fixer --dry-run` but yeah, have to look at the logs to see what went wrong in case of failure.

ekho commented:
@keradus fixed title, rebased on master again

ekho commented:
@keradus up

keradus commented:
@ekho you need to collect some thumbs on PR ;)

ekho commented:
@keradus thumbs? :-O

SpacePossum commented:
LGTM
Just wondering how you are counting number of tests and number of assertions. Wouldn't you want for each time you ran the tool to count as one test ("code style test") with the number of fixers used multiplied by the number of files applied to as assertion count?

ekho commented:
@SpacePossum I thinked about different variants
In my opinion the best variant when we can see files as test cases and applied fixers as failures

keradus commented:
@ekho I'm not using JUnit format anywhere. could you make a build from your fork and show how would that report looks like in any CI you mentioned, just for an example ?

ekho commented:
@keradus in Bamboo it looks like
* [run with invalid file](http://storage4.static.itmages.com/i/16/0413/h_1460553521_2572197_94b20ea4e5.png)
* [summary of run with wrong file](http://storage5.static.itmages.com/i/16/0413/h_1460553838_9399727_8ee10320bc.png)
* [details of run with wrong file](http://storage6.static.itmages.com/i/16/0413/h_1460554136_7629289_42d35a08aa.png)
* [all ok](http://storage5.static.itmages.com/i/16/0413/h_1460554231_4681156_066881e01e.png)

keradus commented on tests/Report/JunitReporterTest.php:
```diff
@@ -0,0 +1,265 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuites>
+  <testsuite name="PHP CS Fixer report" tests="1" assertions="1" failures="1" errors="0">
+    <testcase name="someFile" class="someFile" file="someFile.php" assertions="1">
```
question:
no need for `failures="1"` ?

ekho replied:
`failures` is a number of failed `testcase`s
it should be in the `testsuite` only

keradus commented on tests/Report/JunitReporterTest.php:
```diff
@@ -0,0 +1,265 @@
+  <testsuite name="PHP CS Fixer report" tests="1" assertions="1" failures="1" errors="0">
+    <testcase name="someFile" class="someFile" file="someFile.php" assertions="1">
+      <failure type="csc">Wrong code style</failure>
+      <system-err><![CDATA[Suggested fix:
```
system error ?

ekho replied:
xml schema does not contain another appropriate element =(

But I found that Bamboo does not handle it =(((

keradus commented on tests/Report/JunitReporterTest.php:
```diff
@@ -0,0 +1,265 @@
+        $expectedXml = <<<'XML'
+<?xml version="1.0"?>
+<testsuites>
+  <testsuite assertions="3" errors="0" failures="3" name="PHP CS Fixer report" tests="2" time="1.234">
```
I see no example with errors != 0

ekho replied:
`errors` assumes lint errors (invalid files)
But there is no way to pass lint errors to reporter class

keradus replied:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/src/Console/Command/FixCommand.php#L391-L402

I think we could pass them to reporters if needed

keradus replied:
If not then always having errors=0 is a bit pointless

ekho replied:
Ok, I will update PR.

ekho replied:
https://github.com/junit-team/junit5/blob/master/junit-console/src/main/java/org/junit/gen5/console/tasks/XmlReportWriter.java#L98
always adds attribute `errors` ;)

keradus commented on tests/Report/JunitReporterTest.php:
```diff
@@ -0,0 +1,265 @@
+                        ),
+                    )
+                )
+                ->setMemory(2.5 * 1024 * 1024)
```
where one could see the memory in XML ?

ekho replied:
I do not know =) But it supported by xml schema

keradus replied:
that's my point, you set up time and show it. then you set up memory and not show it

ekho replied:
sorry, I see memory, but think about time =))

May be should keep this to assert that setting memory does not affect xml because schema does not support it (only time) ?

keradus replied:
Question: why scheme doesn't support it ?

keradus commented on tests/Report/JunitReporterTest.php:
```diff
@@ -0,0 +1,265 @@
+<testsuites>
+  <testsuite name="PHP CS Fixer report" tests="1" assertions="1" failures="1" errors="0">
+    <testcase name="someFile" class="someFile" file="someFile.php" assertions="1">
+      <failure type="csc">Wrong code style</failure>
```
what is csc?

ekho replied:
code style conventions =)

keradus replied:
first time seeing the shortcut.

I don't like using it

ekho replied:
I'll change it

keradus commented on src/Report/JunitReporter.php:
```diff
@@ -0,0 +1,145 @@
+        if ($addAppliedFixers) {
+            foreach ($fixResult['appliedFixers'] as $appliedFixer) {
+                $failure = $dom->createElement('failure');
+                $failure->setAttribute('type', $appliedFixer);
```
it's odd that `type` is sometimes name of rule and sometimes magic `csc`

ekho replied:
when `--verbose` then `type` is a fixer name
else we must show whatever, I chose `csc`

keradus replied:
I don't like having one thing with different meaning depending on configuration...

keradus commented on src/Report/JunitReporter.php:
```diff
@@ -0,0 +1,145 @@
+        $testcase = $dom->createElement('testcase');
+        $testcase->setAttribute('name', $testName);
+        $testcase->setAttribute('class', $testName);
+        $testcase->setAttribute('file', $file);
```
why `name`, `class` and `file` ?

`name` === `class`

Also, if we have `file`, what we need `name`/`class` ?

ekho replied:
I added both attrubutes (`name` and `class`) because I do not know how different CI handles this report.
Bamboo interprets dots in `name` and `class` as separator between namespace and class and shows only part after dot (actually it is a file extension - `php`)

keradus replied:
Then, junit isn't standardized ?

ekho replied:
I found some schema versions
* https://svn.jenkins-ci.org/trunk/hudson/dtkit/dtkit-format/dtkit-junit-model/src/main/resources/com/thalesgroup/dtkit/junit/model/xsd/
* https://github.com/gabrielfalcao/lettuce/blob/master/tests/functional/xunit.xsd
* https://windyroad.com.au/dl/Open%20Source/JUnit.xsd ([bamboo uses it](https://confluence.atlassian.com/jirakb/junit-parser-task-in-bamboo-shows-failed-to-parse-test-result-file-error-when-parsing-test-results-file-800701931.html))

Now I think should use latest (http://help.catchsoftware.com/display/ET/JUnit+Format)

keradus replied:
I think that people behind JUnit software are using Travis' one now:
https://github.com/junit-team/junit5/issues/86#issuecomment-176704616

keradus commented on src/Report/JunitReporter.php:
```diff
@@ -0,0 +1,145 @@
+            foreach ($fixResult['appliedFixers'] as $appliedFixer) {
+                $failure = $dom->createElement('failure');
+                $failure->setAttribute('type', $appliedFixer);
+                $appliedFixerReadable = ucfirst(str_replace('_', ' ', $appliedFixer));
```
I would use name of rule as is, without any modifications.
in case of many rules violated it's nice to have the list of them instead of some modified strings

ekho replied:
I will think about modifying report because Bamboo does not support `system-err` element

keradus replied:
If it's possible follow junit general specification, not bamboo one

ekho commented:
@keradus I updated pr

- removed `class` attribute from `testcase`
- only one `failure` element per `testcase` which contains applied fixers and diff (more compatible with different CI tools)

note: used latest Jenkins JUnint xml schema (https://svn.jenkins-ci.org/trunk/hudson/dtkit/dtkit-format/dtkit-junit-model/src/main/resources/com/thalesgroup/dtkit/junit/model/xsd/junit-7.xsd)

SpacePossum commented:
LGTM, but the proof of the pudding is in the eating, so if no other users of JUnit-formatted reports do report issues I'm :+1:   :)

keradus commented on tests/Report/junit.xsd:
```diff
@@ -0,0 +1,104 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
could sb want to use that xsd file?

ekho replied:
sb?

keradus replied:
sorry.
somebody

ekho replied:
For what? They can get it from original location (https://svn.jenkins-ci.org/trunk/hudson/dtkit/dtkit-format/dtkit-junit-model/src/main/resources/com/thalesgroup/dtkit/junit/model/xsd/junit-7.xsd)
I put it here to minimize external dependencies

keradus replied:
for the current state of PR - how could one know which specification are we following ?

ekho replied:
May be should add comment to JunitReporter and README.md ?

keradus replied:
why not

ekho replied:
Could you help me to write right text ? plz :confused:

keradus replied:
next reason to incorporate xsd file in repo.
external specification could be changed. Especially since it's not standard like ECMA-* or sth...

keradus replied:
next thing - this xsd file is subset of one from jenkins or it is exactly the same file ?

ekho replied:
exactly the same file

keradus replied:
And we need it right? So next reason to store it outside tests and point to it instead of external resource.

ekho commented:
@keradus I've added comments as I could

SpacePossum commented on tests/Report/JunitReporterTest.php:
```diff
@@ -0,0 +1,285 @@
+     * Validates generated xml report with schema.
+     * Uses JUnit XML schema from Jenkins.
+     *
+     * @link https://svn.jenkins-ci.org/trunk/hudson/dtkit/dtkit-format/dtkit-junit-model/src/main/resources/com/thalesgroup/dtkit/junit/model/xsd/junit-7.xsd
```
@link is deprecated in PSR5 https://github.com/phpDocumentor/fig-standards/blob/master/proposed/phpdoc.md#710-link-deprecated would you mind changing it to @see?

SpacePossum commented on README.rst:
```diff
@@ -128,7 +128,9 @@ problems as possible on a given file or files in a given directory and its subdi
-The ``--format`` option for the output format. Supported formats are ``txt`` (default one), ``json`` and ``xml``.
+The ``--format`` option for the output format. Supported formats are ``txt`` (default one), ``json``, ``xml`` and ``junit``.
+
+NOTE: When using ``junit`` format report generates in accordance with JUnit xml schema from Jenkins.
```
:+1:

keradus replied:
@SpacePossum what do you think about https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1825#discussion_r59993270 ?

keradus replied:
@ekho should be linkable

SpacePossum replied:
I think it would be best to have all strict output formats (xml, json and junit) formalized using schema's and have those outside of the test scope. 
Maybe start with this one by putting it in a `doc` folder? (and put `contributing` and the `cookbook` there as well?)

keradus replied:
contributing is autolinked in GitHub, I wouldn't move it to `doc`

keradus commented on tests/Report/JunitReporterTest.php:
```diff
@@ -0,0 +1,285 @@
+     * Validates generated xml report with schema.
+     * Uses JUnit XML schema from Jenkins.
+     *
+     * @see https://svn.jenkins-ci.org/trunk/hudson/dtkit/dtkit-format/dtkit-junit-model/src/main/resources/com/thalesgroup/dtkit/junit/model/xsd/junit-7.xsd
```
```html

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL /trunk/hudson/dtkit/dtkit-format/dtkit-junit-model/src/main/resources/com/thalesgroup/dtkit/junit/model/xsd/junit-7.xsd was not found on this server.</p>
<hr>
<address>Apache/2.2.14 (Ubuntu) Server at svn.jenkins-ci.org Port 443</address>
</body></html>
```

ekho replied:
subversion is dead =(

ekho replied:
Can be found here https://github.com/Vlatombe/lib-dtkit-format/blob/master/dtkit-junit-model/src/main/resources/com/thalesgroup/dtkit/junit/model/xsd/junit-7.xsd

keradus replied:
4th reason for https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1825#discussion_r59993270

keradus commented:
Any updates on this one?

ekho commented:
@keradus I'm confused. I do not understand what I should do to finish PR =(((((

SpacePossum commented:
I think there is this one;
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1825#discussion_r59555969

And it would be nice to make a `doc` folder in the root of the project and add the `.xsd` you follow in there. This way the end-users know what the tool generates. We could also move the `cookbook` into there are update links to it to point there. 
It would cool to test the output the tool create against the `.xsd` in tehe doc folder.

Depending on what time https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1910 is done and merged that PR or this one needs updating, I can take care of that.

ekho commented:
svn.jenkins-ci.org died. Actual Jenkins xunit schemas can be found at https://github.com/jenkinsci/xunit-plugin/tree/master/src/main/resources/org/jenkinsci/plugins/xunit/types/model/xsd .
I got latest (junit-10.xsd) and put it to `doc` dir. Test updated to use it.

https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1825#discussion_r59559150
I researched several different xunit schemes. No one supports memory. I think because spent time is very common mark for tests but memory is not (like cpu usage, disk io, etc).

SpacePossum commented on tests/TextDiffTest.php:
```diff
@@ -90,7 +90,7 @@ public function testAllFormatsCovered()
 
         $this->assertSame(
-            array('json', 'txt', 'xml'),
+            array('json', 'junit', 'txt', 'xml'),
```
please add `junit` at the array at line 71 (https://github.com/bankiru/PHP-CS-Fixer/blob/9f2fd37f6f2e8e9eaf495c29ced2548350634106/tests/TextDiffTest.php#L71) as well to test the output (escaping)

ekho replied:
only add `junit` fails test =(

SpacePossum replied:
Which is a good thing, it means we found a little issue before it was merged :)
Please take a look at;
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/src/Report/XmlReporter.php#L71

I think your reporter needs something like that and the test should pass.

ekho commented:
@SpacePossum fixed. Tests fails as well as шт master branch =(

keradus commented:
Damn it :( I have no way to check every possible env during merging one line to another :( At least it's good that Travis detected it.
Please, could you work on a fix for broken master ? (as separated PR)

SpacePossum commented:
The failing test are not related, so no worries.
Could you remove the line commented here; https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1825/files#r59548663?
And maybe move this file;
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/COOKBOOK-FIXERS.md
to the '/doc' folder you created?

I've a look at the failing PSR0/4 tests.

SpacePossum commented:
LGTM :+1:

ekho commented:
offtopic
@keradus @SpacePossum did you think about removing support of old php versions (5.3, 5.4) in the 2.0 release ?

keradus commented:
2.0 release will not dorp 5.3. 2.x may.
too many, to important changes to drop 5.3 without important reason, as 2.x is working on 5.3

keradus commented:
Travis restarted

ekho commented:
May be I should update branch?

keradus commented:
@ekho please rebase
@SpacePossum please review ;)

ekho commented:
done

SpacePossum commented:
Just a few minors https://github.com/bankiru/PHP-CS-Fixer/pull/4

Tested and looks good to me :)

SpacePossum commented:
if the test are green; :ship: :+1:

ekho commented:
@keradus @SpacePossum tests are green =)

keradus commented:
:+1:

keradus commented:
Thank you @ekho.

SpacePossum commented:
thanks @ekho :)
